Add Losses

JimClarke5 opened the pull request:
This PR adds losses to framework. 

All the loss sub-classes inherit from `Loss`.

The `Losses` class, has methods that can be called directly to get raw loss values. These are utilized by the `Loss` subclasses before applying a `Reduction` to the loss. The `Losses` class will also  be used by some of the `Metric` classes when that feature is submitted.

The `impl` package has some helper methods and classes utilized by the `loss` classes, and are not expected to be exposed outside the framework `module`,  when we do modules.

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/impl/LossesImpl.java:
```diff
@@ -0,0 +1,304 @@
+
+public class LossesImpl {
+
+  /**
```
The Javadocs in this file are still partly in markdown.

JimClarke5 replied:
OK, I thought I caught them all, I will fix.

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/impl/LossesImpl.java:
```diff
@@ -0,0 +1,304 @@
+   * @param tf the TensorFlow Ops
+   * @param predictions Predicted values, a <code>Operand</code> of arbitrary dimensions.
+   * @param labels Optional label <code>Operand</code> whose dimensions match <code>prediction</code>.
+   * @return Tuple of <code>prediction</code>, <code>label</code> and <code>sampleWeight</code>. Each of them possibly has the last
```
For this method, the returned `sampleWeight` is always null.

JimClarke5 replied:
That is not always the case when we do Metrics.

deansher replied:
I'm just thinking our documentation for this method might take into account that the returned `sampleWeight` is always `null`.

JimClarke5 replied:
Now I see what you  are talking about. I added a comment in the @return that `sampleWeight` will be `null` for this particular method signature.

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/impl/LossesImpl.java:
```diff
@@ -0,0 +1,304 @@
+   *
+   * @param tf the TensorFlow Ops
+   * @param predictions Predicted values, a <code>Operand</code> of arbitrary dimensions.
+   * @param labels Optional label <code>Operand</code> whose dimensions match <code>prediction
```
Is "match" the right way to describe the precondition relationship between `predictions` and `labels`?

JimClarke5 replied:
It is definitely not the same `Shape`.  I was thinking of compatible, but that has specific meaning in `Shape.isCompatibleWIth`.  The description is saying the ranks must be equal or differ by one. I am not sure of one word that describes that.  `match` was the word used in the Python version of this method.

deansher replied:
Hmm, here's a suggestion:
* We could decide what we want the convention to be, in terms of squeeze-or-expand plus maybe broadcasting.
* Write this up carefully in the class javadoc for either `Loss` or `Losses`.
* Mention that documentation in the class javadoc for every other loss class.
* Also mention it in `Loss#call`.
* And be silent about it in the individual methods of `Losses`.
Perhaps?

JimClarke5 replied:
`match` must mean that the shapes of the input operands are capable of being molded into the relationships defined for the result of this method. Again `LossesImpl` is intended to be marked as module private (JDK 11) and only should be accessible from the losses or metrics package. It is not intended to be a general use API.

Craigacp replied:
We should probably note in the javadoc for the class that this is an internal implementation class and subject to change (and being locked off under the module system).

JimClarke5 replied:
Added this comment for the `LossesImpl` class

```
/**
 * These are helper methods for Losses and will be module private when
 * Java modularity is applied to TensorFlow Java.
 * These methods should not be used outside of the Loss package.
 */
```

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/impl/LossesImpl.java:
```diff
@@ -0,0 +1,304 @@
+
+    if (labels != null) {
+      Shape labelsShape = labels.asOutput().shape();
+      long labelRank = labelsShape.numDimensions();
```
For consistency, `labelsRank`.

JimClarke5 replied:
OK

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/impl/LossesImpl.java:
```diff
@@ -0,0 +1,304 @@
+      long labelRank = labelsShape.numDimensions();
+      if (labelRank != Shape.UNKNOWN_SIZE && predictionsRank != Shape.UNKNOWN_SIZE) {
+        // Use static rank for `label` and `prediction`.
+        if (predictionsRank - labelRank != 1 || predictionsShape.size(-1) == 1) {
```
This doesn't seem to do what the documentation says it will do? My domain knowledge is so weak that I hate to claim it's wrong, but the `!=` and the fact that we do nothing if this second condition fails are certainly surprising.

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/impl/LossesImpl.java:
```diff
@@ -0,0 +1,304 @@
+  public static <T extends TNumber> Tuple<T> removeSqueezableDimensions(
+      Ops tf, Operand<T> labels, Operand<T> predictions, int expectedRankDiff) {
+
+    tf = tf.withSubScope("removeSqueezableDimensions");
```
We don't open sub-scopes very often. Should we do it more? Or why here?

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/impl/LossesImpl.java:
```diff
@@ -0,0 +1,304 @@
+   * Squeeze last dim if ranks differ from expected by exactly 1.
+   *
+   * @param tf the TensorFlowOps
+   * @param labels Label values, a `Tensor` whose dimensions match `predictions`.
```
"match" seems at best vague here.

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java:
```diff
@@ -0,0 +1,683 @@
+   */
+  public static <T extends TNumber, U extends TNumber> Operand<T> meanAbsoluteError(
+      Ops tf, Operand<U> labels, Operand<T> predictions) {
+    Operand<T> tLabels = tf.dtypes.cast(labels, predictions.asOutput().dataType());
```
Do we want to avoid this cast in the case where `labels` already has the same data type?

JimClarke5 replied:
I guess the question is what is the overhead of casting onto oneself vs the overhead of checking?  I would hope that `tf.dtypes.cast` already handles this, but I could be mistaken.

deansher replied:
The code for checking could be something like this:
~~~ Java
@SuppressWarnings("unchecked")
private static <T extends TNumber, U extends TNumber> Operand<T> castIfNecessary(
    Operand<U> value, DataType<T> requiredType) {
  return (value.asOutput().dataType() == requiredType) 
      ? (Operand<T>) value
      : tf.dtypes.cast(value, requiredType);
}
~~~
So the overhead of checking would be the function call plus `value.asOutput().dataType() == requiredType`.

Looking at the code for `tf.dtypes.cast`, unless we think a cast is almost always needed, it would be cheaper to do the check to sometimes avoid it.
~~~ Java
  public <U extends TType, T extends TType> Cast<U> cast(Operand<T> x, DataType<U> DstT,
      Cast.Options... options) {
    return Cast.create(scope, x, DstT, options);
  }

  @Endpoint(describeByClass = true)
  public static <U extends TType, T extends TType> Cast<U> create(Scope scope, Operand<T> x, DataType<U> DstT, Options... options) {
    OperationBuilder opBuilder = scope.env().opBuilder("Cast", scope.makeOpName("Cast"));
    opBuilder.addInput(x.asOutput());
    opBuilder = scope.applyControlDependencies(opBuilder);
    opBuilder.setAttr("DstT", DstT);
    if (options != null) {
      for (Options opts : options) {
        if (opts.Truncate != null) {
          opBuilder.setAttr("Truncate", opts.Truncate);
        }
      }
    }
    return new Cast<U>(opBuilder.build());
  }
~~~

JimClarke5 replied:
I like `castIfNecessary` as a general util method. It would be used almost everywhere, so it would be a huge change.
Perhaps create a new PR for `castIfNecessary`, then once that is merged we can start retrofitting all packages under framework.

deansher replied:
In graph construction mode, an unnecessary call to `cast` creates an unnecessary graph operation.

Craigacp replied:
*shrug* it'll be a no-op most of the time and compiled away if we get XLA working. Given the relative size of the computation around it I suspect it won't be an issue.

karllessard replied:
I also vote for a explicit check in the code to avoid adding an extra operation to the graph when it is not required

JimClarke5 replied:
OK, I will add a helper class in` org.tensorflow.framework.utils`, then retrofit the `Loss` classes.

JimClarke5 replied:
Just a comment on @deansher proposed method here, the datatypes for `<U>` and `<T>` should not be restricted to `TNumber` because it is valid to cast to/from `TNumber` and `TBool` .

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java:
```diff
@@ -0,0 +1,683 @@
+  public static <T extends TNumber, U extends TNumber> Operand<T> meanAbsolutePercentageError(
+      Ops tf, Operand<U> labels, Operand<T> predictions) {
+    DataType<T> dataType = predictions.asOutput().dataType();
+    Operand<T> tLabels = tf.dtypes.cast(labels, predictions.asOutput().dataType());
```
Can just use `dataType`.

JimClarke5 replied:
Fixed

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java:
```diff
@@ -0,0 +1,683 @@
+   * @param <T> the data type of the Operands
+   * @return the binary crossentropy loss.
+   */
+  private static <T extends TNumber> Operand<T> binaryCrossentropy(
```
I tripped over this private method having the usual naming of a loss method, since I didn't notice that it was private and so expected it to follow the conventions of public loss methods, such as invoking `squeezeOrExpandDimensions`.  Also (if I'm navigating through unfamiliar territory accurately), this method doesn't compute a `binaryCrossentropy` since it is depending on its caller to compute the mean at the end.

JimClarke5 replied:
This method does the grunt work for the binaryCrossentropy after the operands have had their shapes and types manipulated and after smoothing the labels. Perhaps a new name would remove some of the confusion.

deansher replied:
Yes, I wonder if we want to call it something like `binaryCrossentropyHelper`?

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/CategoricalHinge.java:
```diff
@@ -0,0 +1,91 @@
+import org.tensorflow.types.family.TNumber;
+
+/**
+ * Computes the categorical hinge loss between labels and predictions.
```
Should we follow the Python in documenting that labels are expected to be `0` or `1`?

JimClarke5 replied:
Yes, The Python `CategporicalHinge` class does not mention that at all, but it is mentioned in the `loss_hinge`.
method.

I have added an entry to the class JavaDoc and to the `Losses.categoricalHinge` method.

JimClarke5 replied:
Actually the values can be` [-1, 0, 1]`.` [0,1]` is converted to `[-1,1`]. I have added a value check to make sure the values are wholly contained in the allowed values set ` [-1, 0, 1]`. This will either throw `TFInvalidArgumentException` if run in Graph mode via a control dependency, and throw `IllegalArgumentException` if created in Eager mode with the `call` method.

deansher replied:
Cool -- Resolved.

Craigacp replied:
What does it do if there are all three of `[-1 0 1]` present? That's probably an invalid input, does it throw?

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/BinaryCrossentropy.java:
```diff
@@ -0,0 +1,179 @@
+  /**
+   * Creates a Binary Crossentropy Loss using {@link Class#getSimpleName()} as the loss name, {@link
+   * #FROM_LOGITS_DEFAULT} for fromLogits, {@link #LABEL_SMOOTHING_DEFAULT} for labelSmoothing and a
+   * Loss Reduction of {@link * Reduction#AUTO}
```
Deleted

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/BinaryCrossentropy.java:
```diff
@@ -0,0 +1,179 @@
+      Operand<U> labels, Operand<T> predictions, Operand<T> sampleWeights) {
+    Operand<T> losses =
+        Losses.binaryCrossentropy(tf, labels, predictions, fromLogits, labelSmoothing);
+    return LossesImpl.computeWeightedLoss(tf, losses, getReduction(), sampleWeights);
```
Changed to getTF()

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/CategoricalCrossentropy.java:
```diff
@@ -0,0 +1,219 @@
+  /**
+   * Creates a categorical cross entropy Loss using {@link Class#getSimpleName()} as the loss name,
+   * {@link #FROM_LOGITS_DEFAULT} for fromLogits, {@link #LABEL_SMOOTHING_DEFAULT} for
+   * labelSmoothing, a Loss Reduction of {@link * Reduction#AUTO}, and an axis of {@link
```
Removed all Extraneous `@link *`

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/CategoricalCrossentropy.java:
```diff
@@ -0,0 +1,219 @@
+          Operand<U> labels, Operand<T> predictions, Operand<T> sampleWeights) {
+    Operand<T> losses =
+        Losses.categoricalCrossentropy(tf, labels, predictions, fromLogits, labelSmoothing, axis);
+    return LossesImpl.computeWeightedLoss(tf, losses, getReduction(), sampleWeights);
```
Changed to `getTF()`

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Loss.java:
```diff
@@ -0,0 +1,91 @@
+   *
+   * @param labels the truth values or labels
+   * @param predictions the predictions
+   * @param <T> The data type of the labels, predictions and loss.
```
Fixed

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Loss.java:
```diff
@@ -0,0 +1,91 @@
+   *     predictions is scaled by the corresponding value of sample_weight. (Note on dN-1: all loss
+   *     functions reduce by 1 dimension, usually axis=-1.)
+   * @param <T> The data type of the predictions, sampleWeights and loss.
+   * @param <U> The data type of the labels.
```
The standard convention is:

- E - Element (used extensively by the Java Collections Framework)
- K - Key
- N - Number
- T - Type
- V - Value
- S,U,V etc. - 2nd, 3rd, 4th types

[Generic Types](https://docs.oracle.com/javase/tutorial/java/generics/types.html)

deansher replied:
I don't think of that list as being especially prescriptive. In the Oracle documentation you link above, the list is introduced as "The most commonly used type parameter names are: ...".

[Google's Java style guide](https://google.github.io/styleguide/javaguide.html#s5.2.8-type-variable-names) says:
> Each type variable is named in one of two styles:
> 
> * A single capital letter, optionally followed by a single numeral (such as E, T, X, T2)
> * A name in the form used for classes (see Section 5.2.2, Class names), followed by the capital letter T (examples: RequestT, FooBarT).

Personally, I'd lean toward using some of our own single-letter conventions for situations that are common in our own code, including `L` as the labels type.

KartikChugh replied:
> Personally, I'd lean toward using some of our own single-letter conventions for situations that are common in our own code, including `L` as the labels type.

This may be hard to follow consistently once several letters have been used e.g. 'L' might be needed for something other than label type. Seems a tad more confusing than the standard type names

deansher replied:
Ok, sticking with the original plan! Resolved.

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java:
```diff
@@ -0,0 +1,683 @@
+   * @param tf The TensorFlow Ops
+   * @param labels the labels
+   * @param predictions the predictions
+   * @param <T> the data type of the result
```
OK

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/impl/Tuple.java:
```diff
@@ -0,0 +1,53 @@
+ *
+ * @param <T> the data type of the Tuple entries.
+ */
+public class Tuple<T extends TNumber> {
```
This object will also be used in Metrics as many metrics are built using loss classes or `Losses` methods. I have changed it to `LossTuple`.

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Loss.java:
```diff
@@ -0,0 +1,91 @@
+   * Creates a Loss using a Loss Reduction of {@link Reduction#AUTO}
+   *
+   * @param tf the TensorFlow Ops
+   * @param name the name of this Loss
```
Why would someone want to pass `null`, when there are other CTORs that handle that condition?

deansher replied:
For APIs that will get enough use to be worth some polish, I tend toward carefully documenting edge cases. I don't know whether we want to invest in that now.

JimClarke5 replied:
OK, added this to `name` param, `if null the name will be {@link Class#getSimpleName()}.`

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/impl/LossesImpl.java:
```diff
@@ -0,0 +1,304 @@
+   * @param <T> the data type of the loss
+   * @return the weighted loss
+   */
+  public static <T extends TNumber> Operand<T> computeWeightedLoss(
```
This is another case where the combination of `squeezeOrExpandDimensions` and broadcasting yields a complex relationship between shapes of `sampleWeight`, `loss`, and the return value. In particular, due to broadcasting in `mul`, in the case of `reduction` == `NONE` and a surprising shape of `sampleWeight`, the return value may have a very different shape than `loss`.

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/impl/LossesImpl.java:
```diff
@@ -0,0 +1,304 @@
+   */
+  private static <T extends TNumber> int[] allAxis(Operand<T> op) {
+    int rank = op.asOutput().shape().numDimensions();
+    int[] axes = new int[rank];
```
`rank` could be `-1` at this point.

JimClarke5 replied:
Fixed

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/impl/LossesImpl.java:
```diff
@@ -0,0 +1,304 @@
+   * @param <T> the type of Operand
+   * @return the integer array representing all the axes of the operand.
+   */
+  private static <T extends TNumber> int[] allAxis(Operand<T> op) {
```
`allAxes`?

JimClarke5 replied:
Changed name to allAxes

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/impl/LossesImpl.java:
```diff
@@ -0,0 +1,304 @@
+    }
+    Shape weightsShape = sampleWeight.asOutput().shape();
+    long weightsRank = weightsShape.numDimensions();
+    if (weightsRank == 0) { // scalar
```
What should happen if `weightsRank` is `UNKNOWN`?

JimClarke5 replied:
It falls through and executes the last part of the method after the `// Use dynamic rank.` comment.

deansher replied:
:-) Oh yeah, that.

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/impl/LossesImpl.java:
```diff
@@ -0,0 +1,304 @@
+
+    if (predictionsRank != Shape.UNKNOWN_SIZE && weightsRank != Shape.UNKNOWN_SIZE) {
+
+      if (weightsRank - predictionsRank == 1) {
```
Here we are working with the original `predictionsRank`, when we wanted to be working with the new rank.

JimClarke5 replied:
This matches the original Python code, but when you think about it, the `predictions` rank would never change from UNKNOWN to KNOWN and vice versa in a static context.

deansher replied:
I was thinking perhaps `predictions` changed rank through our squeezing it to match `labels` earlier in the method. But I think there's a more pernicious problem. Here's an elided version of some of the method's code. Notice that we may squeeze `predictions`, but we only store the result in `tuple`. If we then also work with `sampleWeight`, we neither reference the squeezed version of `predictions` nor return it.
~~~ Java
    if (labels != null) {
        . . . 
        if (predictionsRank - labelRank != 1 || predictionsShape.size(-1) == 1) {
          tuple = removeSqueezableDimensions(tf, labels, predictions);
        }
      } else { // use dynamic rank
        tuple = removeSqueezableDimensions(tf, labels, predictions);
      }
    }
    . . .
    if (predictionsRank != Shape.UNKNOWN_SIZE && weightsRank != Shape.UNKNOWN_SIZE) {

      if (weightsRank - predictionsRank == 1) {
        sampleWeight = tf.squeeze(sampleWeight);
        . . .
      }
      return new Tuple<>(labels, predictions, sampleWeight);
    }
~~~

JimClarke5 replied:
Yes, we should probably fetch the `labels` and `predictions` from `tuple` first. I'll fix it.

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/impl/LossesImpl.java:
```diff
@@ -0,0 +1,304 @@
+   *     Each of them possibly has the last dimension squeezed, <code>sampleWeight</code> could be
+   *     extended by one dimension. If <code>sampleWeight</code> is null, (prediction, label) is
+   *     returned.
+   */
```
This method has a myriad of complex cases, so I think it deserves its own direct unit test.

JimClarke5 replied:
I could not find direct test cases for this method in Python. It's defined in `tensorflow/tensorflow/python/ops/losses/utils.py`. You want to take a stab at it?

deansher replied:
:-) Totally. I want to do some work on #92 first, so I'll open an issue for myself.

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java:
```diff
@@ -0,0 +1,683 @@
+  public static <T extends TNumber, U extends TNumber> Operand<T> meanSquaredLogarithmicError(
+      Ops tf, Operand<U> labels, Operand<T> predictions) {
+    DataType<T> dataType = predictions.asOutput().dataType();
+    Operand<T> tLabels = tf.dtypes.cast(labels, predictions.asOutput().dataType());
```
Fixed, hopefully I have fixed them all.

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java:
```diff
@@ -0,0 +1,683 @@
+  }
+
+  /**
+   * Calculates the mean squared logarithmic percentage error between labels and predictions.
```
Fixed

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/impl/LossesImpl.java:
```diff
@@ -0,0 +1,304 @@
+    // Use dynamic rank.
+
+    // TODO Operand<TInt32> rankDiff = tf.math.sub(tf.rank(predictions), tf.rank(labels));
+    if (predictionsRank == Shape.UNKNOWN_SIZE && Shape.isCompatible(predictionsShape.size(-1), 1)) {
```
Correct, it should have been `or` not `and`.

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/impl/LossesImpl.java:
```diff
@@ -0,0 +1,304 @@
+    if (labels != null) {
+      Shape labelsShape = labels.asOutput().shape();
+      long labelRank = labelsShape.numDimensions();
+      if (labelRank != Shape.UNKNOWN_SIZE && predictionsRank != Shape.UNKNOWN_SIZE) {
```
This logic is checking to see if both objects ranks are known (not` Shape.unknown()`). If both ranks are known, then it checks to see if the shapes are already in the right relationship or not. If not in the right relationship, then call `removeSqueezableDimensions`. It is basically an optimization to avoid doing the work in `removeSqueezableDimensions` if it does not need to be done.

karllessard commented:
Hi @JimClarke5 , what would be the best order for reviewing your PRs? You have this one, #123 and #106 that are still opened.

JimClarke5 commented:
I would do activations #123 first.

JimClarke5 commented:
@karllessard I have started working on Metrics which depends on Loss, and thusly this PR. I plan to do Metrics in two PRs, the first PR will focus on Metrics that depend on `Losses` functions. The second PR will focus on Metrics that don't depend on `Losses` functions, so these are more complicated to implement. My goal is to set up the scaffolding in the first Metric PR, then focus on the functionality required in the second PR.  As to which existing PR to work on after `activations` #123 , may I suggest that this PR will need to be done before I create the first Metric PR.

Craigacp commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java:
```diff
@@ -0,0 +1,685 @@
+        // TODO  if( output.op().numOutputs() != 1)
+        //          throw new IllegalArgumentException("output can only have 1 output");
+        // TODO output = output.op.inputs[0]
+        fromLogits = true;
```
But it didn't extract the logits, so won't this perform the wrong calculation?

JimClarke5 replied:
Correct, I really need the inputs to the `Softmax` Operation. This is the same issue as mentioned before, 
but I correctly marked these as TODO's

Craigacp replied:
Yeah, but I think it would be best to just ignore this optimization and leave the TODO in, rather than half put it in and have it break in the common case.

JimClarke5 replied:
I commented out this code and marked it as a TODO.

Craigacp commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java:
```diff
@@ -0,0 +1,685 @@
+   * @param labels true targets
+   * @param labelSmoothing A number in the range [0, 1]. When 0, no smoothing occurs. When &gt; 0,
+   *     compute the loss between the predicted labels and a smoothed version of the true labels,
+   *     where the smoothing squeezes the labels towards 0.5. Larger values of labelSmoothing
```
The doc is wrong here, it scales towards `1/n`.

JimClarke5 replied:
Fixed the doc to:

 ```
 * @param labelSmoothing Float in <code>[0, 1]</code>. When <code>&gt; 0</code>, label values are smoothed, meaning the
   *    confidence on label values are relaxed. e.g. <code>label_smoothing=0.2<code> means that we will use a
   *    value of </code>0.1<code> for label </code>0<code> and </code>0.9<code> for label </code>1<code>
```

Craigacp commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/SparseCategoricalCrossentropy.java:
```diff
@@ -0,0 +1,170 @@
+
+  /**
+   * Creates a SparseCategoricalCrossentropy loss using {@link Class#getSimpleName()} as the loss
+   * name, a Loss Reduction of {@link Reduction#AUTO}, and fromLogits={@link #FROM_LOGITS_DEFAULT}.
```
This links to `Reduction#AUTO` instead of `REDUCTION_DEFAULT`.

JimClarke5 replied:
Fixed

Craigacp commented on tensorflow-framework/src/test/java/org/tensorflow/framework/losses/CategoricalCrossentropyTest.java:
```diff
@@ -0,0 +1,213 @@
+        float[] predArray = {
+          1.f, 0.f, 0.f,
+          0.f, 1.f, 0.f,
+          0.f, 0.f, 1.F
```
The last `f` is in caps. Is that intentional? It's not consistent throughout the file if so.

JimClarke5 replied:
Changed all to `F`

Craigacp reviewed (changes requested):
In addition to the specific comments, I think it might be a good idea to add checks that the values are in the expected range (i.e. if it's expecting probabilities then it should check that they are in the range 0-1). Otherwise it's a right pain to track that down. Not sure if it will add too much overhead, but the loss computation tends to be much cheaper than the forward or backward passes, so hopefully it'll be fine.

JimClarke5 commented:
@Craigacp we could set a range check, but it would have to be a control dependency e.g. using `tf.assertThat`

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Loss.java:
```diff
@@ -0,0 +1,91 @@
+   * Creates a Loss
+   *
+   * @param tf the TensorFlow Ops
+   * @param name the name of this loss
```
OK, added this to all  `name` param,` if null the name will be {@link Class#getSimpleName()}.`

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/CategoricalCrossentropy.java:
```diff
@@ -0,0 +1,219 @@
+  /**
+   * Creates a categorical cross entropy Loss using {@link Class#getSimpleName()} as the loss name,
+   * {@link #FROM_LOGITS_DEFAULT} for fromLogits, {@link #LABEL_SMOOTHING_DEFAULT} for
+   * labelSmoothing, a Loss Reduction of {@link Reduction#AUTO}, and an axis of {@link
```
I fixed all instances across all classes  to use `Loss#REDUCTION_DEFAULT`  which is set to `Reduction.AUTO` in the `Loss` class.

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/CategoricalCrossentropy.java:
```diff
@@ -0,0 +1,219 @@
+   *
+   * @param tf the TensorFlow Ops
+   * @param fromLogits Whether to interpret predictions as a tensor of logit values
+   * @param labelSmoothing Float in [0, 1]. When 0, no smoothing occurs. When > 0, we compute the
```
The formula for label smoothing is, ( labelSmoothing != 0 does no smoothing so this logic is skipped in that case).

`labels` * (1-`lableSmoothing`)+ (`labelSmoothing` * 0.5)

So if labelSmoothing = 1, then (labels * 0 + 0.5) = 0.5. So, when labelSmoothing == 1, the labels are a constant 0.5.

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/LogCosh.java:
```diff
@@ -0,0 +1,99 @@
+  }
+
+  /**
+   * Creates a Kullback Leibler Divergence Loss
```
Copy & paste error, I fixed it.

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java:
```diff
@@ -0,0 +1,685 @@
+   * @param fromLogits Whether to interpret predictions as a tensor of logit values
+   * @param labelSmoothing Float in [0, 1]. When 0, no smoothing occurs. When &gt; 0, compute the
+   *     loss between the predicted labels and a smoothed version of the true labels, where the
+   *     smoothing squeezes the labels towards 0.5. Larger values of labelSmoothing correspond to
```
Actually this documentation belongs to BinaryCrossentropy, 
Here is what it should be:

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java:
```diff
@@ -0,0 +1,685 @@
+        if (predictions.op().numOutputs() != 1)
+          throw new IllegalArgumentException("output can only have 1 output");
+        predictions = predictions.op().output(0);
+        return tf.nn.softmaxCrossEntropyWithLogits(tLabels, predictions, -1);
```
This was a TODO and the output() what just  a holder until I figured out how to get the inputs to the
`SoftMax` operation. If you don't have a suggestion, maybe remove this logic for now.

Here is the logic from Python.

```
    # When softmax activation function is used for output operation, we
    # use logits from the softmax function directly to compute loss in order
    # to prevent collapsing zero when training.
    # See b/117284466
    assert len(output.op.inputs) == 1
    output = output.op.inputs[0]
    return nn.softmax_cross_entropy_with_logits_v2(
        labels=target, logits=output, axis=axis)
```

JimClarke5 replied:
I commented out this code and marked it as a TODO.

JimClarke5 commented:
I tested the range check, but found out that `Eager` mode does not support Control Dependencies.  I have modified my function to check if its `Graph` or `Eager`. `Graph` mode produces a Control Dependency, while `Eager` mode throws `IllegalArgumentException`. The new method is in `LossesImpl.rangeCheck`. I have to fix the rest of the test cases and see what other Classes besides xxxCrossentropy need this feature.

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java:
```diff
@@ -0,0 +1,685 @@
+   *
+   * <p>Note that it is a number between -1 and 1. When it is a negative number between -1 and 0, 0
+   * indicates orthogonality and values closer to -1 indicate greater similarity. The values closer
+   * to 1 indicate greater dissimilarity. This makes it usable as a loss function in a setting where
```
The JavaDoc on the class should have matched the JavaDoc on `Losses.cosineSimilarity` method which returns values between -1 and 1. I have fixed the JavaDoc on the class `CosineSimilarity` to match the method JavaDoc.

JimClarke5 replied:
There is a discrepancy between the Python doc for `CosineSimilarity ` and the loss function `cosine_similarity`. Though the doc on `CosineSimilarity` indicates a value in the range of `[-1 to  0]`, the actual method indicates a range of `[-1 to  1]`.
There is nothing in the code between the two, that acts differently as `CosineSimilarity` just calls `cosine_similarity`.

The original `Keras` documentation on `CosineSimilarity`, agrees with the` [-1  to  1]` range. I am guessing the TensorFlow `CosineSimilarity` documentation is wrong.

Another weirdness creeps in in `metrics.CosineSimilarity`, which ends up calling another function called `cosine_proximity`. While the two different functions are in a similar category, they actually compute differently.
We'll address that issue in the `metrics` PR, as perhaps the metrics class should be renamed to `CosineProximity`.

Craigacp replied:
We should probably open an issue on the main TF repo flagging the docs issue, but that can wait.

Craigacp replied:
This javadoc is better, but I think it should mention that this function is inverted from the regular cosine similarity, as that's 1 when the values are most similar and -1 when they point in opposite directions. It makes sense that it is inverted because then you can minimise it sensibly, but it is confusing if you're just browsing through.

JimClarke5 commented:
I have added 2 methods to `LossesImpl`, `rangeCheck` and `valueCheck`.
`rangeCheck` checks that all the values are within the provided min and max inclusive, eg `[0. - 1.]`.
`valueCheck` will check that all the values are contained in a set of values (e.g.` [-1, 0, 1]`).

FIrst, control dependancies do not work in Eager mode. To handle this, I throw an `IllegalArgumentException` from the call method when in Eager mode if the check fails. In Graph mode, I set up a control dependency for the check, and it it will throw `TFInvalidArgumentException` if the check fails  when the loss `Operand` is run.

One question, should these utilities be stored in a common location like `framework.utils`.  I suspect that metrics will have the same checks, but metrics could easily still call the `LossesImpl` methods.

karllessard commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java:
```diff
@@ -0,0 +1,711 @@
+import org.tensorflow.types.family.TNumber;
+
+/** Built-in loss functions. */
+public class Losses {
```
What is the reason for having a seperate `Losses` class that implements all individual loss algorithms instead of implementing them directly in their respective class?

I guess is to offer the choice to the user to choose an object-oriented API or a functional one? Then if we offer this, should we do the same for all other concepts in the framework?

I wondering if we shouldn't take that decision for our users and only present a single API to accomplish a given task, for simplicity and consistency. I personally have a preference for the OO approach.

JimClarke5 replied:
Metrics class will have to call these methods directly. The handling of the `reduction` is different in metrics from losses.

JimClarke5 replied:
Also, the methods in `Losses`, do not apply a `reduction`. `Loss` and `Metric` handle the reduction differently, so the raw `loss` needs to be fetched first. If we didn't have the `loss` methods , the the metric classes would have to duplicate that functionality. Also, I can only assume that Keras has this method split from the class for some reason. Of course, we could move the static method to its corresponding class and leave it static so that it can be still invoked independent of any reduction logic.

It is worth noting, that not all `Metrics` are based on `Loss` algorithms, such as `AUC` (Area under the curve).

Also, there is one class,
`CosineSimilarity`, that uses a different algorithm between the Loss class (cosineSimilarity) as opposed to the corresponding Metric class (cosineProximity). While these two methods are in the same category, they do not compute the exact same result. I tried to reasearch why there is a difference, maybe @Craigacp can shed some light on this. One of the questions in `Metrics` is should the metric class be renamed to CosineProximity.

Craigacp replied:
That's because the losses version of cosine similarity isn't actually cosine similarity, it's the negative of it. Because the cosine of 0 is 1, but here when the two vectors are pointing in the same direction it emits -1. The metrics one actually computes cosine similarity.

I assume it's to make the minimize function work consistently.

karllessard commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/impl/LossesImpl.java:
```diff
@@ -0,0 +1,402 @@
+ * These are helper methods for Losses and will be module private when Java modularity is applied to
+ * TensorFlow Java. These methods should not be used outside of the Loss package.
+ */
+public class LossesImpl {
```
For me, a `*Impl` should be the implementation of an interface, this one looks more like a `LossesHelper` with all its static methods (and the class should probably be final).

I did not went through the whole thing but it looks like these helpers could also be moved directly to `Loss` as protected methods?

JimClarke5 replied:
The split really comes for module visibility. `Losses` should be publicly accessible, while `LossesImpl` should be module private. Some `LossesImpl` methods may be used by `metrics`. Whether we call it `LossesImpl` of `LossesHelper` is a matter of preference. The current methods in `LossesImpl` should not be restricted to Loss classes as `metrics` classes may also make use of them, therefore `protected` is not the right semantic, .

deansher replied:
It feels uncomfortable to me that we plan to use the `LossesImpl` methods from other parts of our framework while restricting them from public use. When a system's built-ins rely on privileged capabilities that aren't available to 3rd-party code, I think it is commonly a big problem for the system's extensibility. In this case, I do see room to argue that these methods aren't "capabilities", but are just "implementation" which can safely be hidden. But given that it is important to us to reuse them for our own metrics, I lean toward thinking of them as capabilities that we should expose.

JimClarke5 replied:
There is a tight symmetry between Losses and Metrics as many (but not all) metrics rely on the methods in Losses.
Don't think other packages will have this close of a relationship.

KartikChugh replied:
Is there a potential use case justifying exposing these to the public? Seeing as they are utilities needed to implement Losses/Metrics. 

Agree with a rename to LossesHelper or LossesUtility to differentiate from interface implementation, however.

JimClarke5 commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/CategoricalHinge.java:
```diff
@@ -0,0 +1,93 @@
+ * <p><code>loss = maximum(neg - pos + 1, 0)</code> where <code>neg=maximum((1-labels)*predictions)
+ * </code> and <code>pos=sum(labels*predictions)</code>
+ *
+ * <p><code>labels</code> values are expected to be 0 or 1.</p>
```
It doesn't throw an exception and the Python test case uses multiple values set to one. There is a doc discrepancy between the documentation for CategorialHinge and the method for categorical_hinge wrt to expecting the values to be either 0 and 1, in both PythonTF an standalone Keras.

KartikChugh commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/impl/LossesImpl.java:
```diff
@@ -0,0 +1,303 @@
+import org.tensorflow.types.family.TNumber;
+
+import java.util.Collections;
+
```
Class should have Javadoc description, no?

JimClarke5 replied:
I have changed the class name to `LossesHelper`.

I don't understand your comment on Class JavaDoc. This is what I have in my copy.
```
/**
 * These are helper methods for Losses and Metrics and will be module private when Java modularity is applied to
 * TensorFlow Java. These methods should not be used outside of the losses and metrics packages.
 */
```

The basic comment was put in a while a ago, and I just updated it to mention `metrics`.

KartikChugh replied:
Rendering issue I think. Looks good, thanks.

karllessard commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java:
```diff
@@ -0,0 +1,728 @@
+  //   * @param <T> The data type of the predictions, sampleWeights and loss.
+  //   * @param <U> The data type of the labels.
+  //   * @return the loss
+  //   *
```
Can we remove this commented-out documentation?

JimClarke5 replied:
OK

karllessard commented:
@Craigacp , @deansher , can you please go through all your previous comments/suggestions/questions and mark as resolved those that are either resolved by a fix from @JimClarke5 or are minor and can wait? The PR looks mature enough to be merged soon.

JimClarke5 commented:
@karllessard  Build failing after push. 

`Error:  Failed to execute goal on project tensorflow-core-api: Could not resolve dependencies for project org.tensorflow:tensorflow-core-api:jar:0.3.0-SNAPSHOT: Could not find artifact org.tensorflow:tensorflow-core-api:jar:linux-x86_64:0.3.0-SNAPSHOT in ossrh-snapshots (https://oss.sonatype.org/content/repositories/snapshots) -> [Help 1]`

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java:
```diff
@@ -0,0 +1,683 @@
+   *
+   * @param tf The TensorFlow Ops
+   * @param labels the labels
+   * @param predictions the predictions
```
We can return to the "`squeezeOrExpandDimensions` followed by broadcasting" topic when I work on #130 .

Resolved.

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java:
```diff
@@ -0,0 +1,683 @@
+    predictions = ops.getTarget();
+    tLabels = ops.getLabels();
+    return tf.math.mean(
+        tf.math.abs(tf.math.sub(tLabels, predictions)), tf.constant(-1), Mean.keepDims(false));
```
We can return to the "`squeezeOrExpandDimensions` followed by broadcasting" topic when I work on #130 .

Resolved.

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java:
```diff
@@ -0,0 +1,683 @@
+    predictions = ops.getTarget();
+    tLabels = ops.getLabels();
+    return tf.math.mean(
+        tf.math.abs(tf.math.sub(tLabels, predictions)), tf.constant(-1), Mean.keepDims(false));
```
minor -- we'll call it Resolved.

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java:
```diff
@@ -0,0 +1,683 @@
+   * Compute binary crossentropy loss between labels and predictions.
+   *
+   * @param tf the TensorFlow Ops
+   * @param target the target Operand
```
minor -- we'll call it Resolved.

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java:
```diff
@@ -0,0 +1,683 @@
+  private static <T extends TNumber> Operand<T> binaryCrossentropy(
+      Ops tf, Operand<T> target, Operand<T> output, boolean fromLogits) {
+    if (fromLogits) {
+      return tf.nn.sigmoidCrossEntropyWithLogits(target, output);
```
We can return to the "`squeezeOrExpandDimensions` followed by broadcasting" topic when I work on #130 .

Resolved.

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java:
```diff
@@ -0,0 +1,683 @@
+      tLabels = smoothLabelsCatX(tf, tLabels, labelSmoothing);
+    }
+    if (fromLogits) {
+      return tf.nn.softmaxCrossEntropyWithLogits(tLabels, predictions, -1);
```
We can return to the "`squeezeOrExpandDimensions` followed by broadcasting" topic when I work on #130 .

Resolved.

deansher commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java:
```diff
@@ -0,0 +1,683 @@
+    // Compute cross entropy from probabilities.
+    Operand<T> cce =
+        tf.reduceSum(
+            tf.math.mul(tLabels, tf.math.log(predictions)),
```
We can return to the "`squeezeOrExpandDimensions` followed by broadcasting" topic when I work on #130 .

Resolved.

deansher reviewed (approved):
I feel great about merging this as it stands! It's a huge and excellent chunk of work.

I don't think I can literally mark my comments as resolved since I don't have write access, but I replied to my comments stating they are resolved and explaining why.

saudet commented:
> @karllessard Build failing after push.

That's just a transient network error:
```
2020-11-12T15:19:41.5511149Z [WARNING] Could not transfer metadata org.tensorflow:tensorflow-core-api:0.3.0-SNAPSHOT/maven-metadata.xml from/to ossrh-snapshots (https://oss.sonatype.org/content/repositories/snapshots): Transfer failed for https://oss.sonatype.org/content/repositories/snapshots/org/tensorflow/tensorflow-core-api/0.3.0-SNAPSHOT/maven-metadata.xml 502 Bad Gateway
```

karllessard commented:
@saudet , @JimClarke5 , it might also be related that we are having trouble these days to build our artifacts as you can [see here](https://github.com/tensorflow/java/actions/runs/337757118), the linux GPU build runs out of space and that prevent the last step to occur (i.e. the bulk deploy that normalize all snapshots timestamp which might explain why a few artifacts disappeared lately: #142).

I’ve retried many times but without success. Samuel, any idea how to solve this again?

saudet commented:
> I’ve retried many times but without success. Samuel, any idea how to solve this again?

It just looks like GitHub Actions is below its guaranteed 14 GB of disk space:
```
2020-11-12T04:02:15.8038646Z Filesystem      Size  Used Avail Use% Mounted on
2020-11-12T04:02:15.8039270Z overlay          84G   75G  8.9G  90% /
2020-11-12T04:02:15.8039572Z tmpfs            64M     0   64M   0% /dev
2020-11-12T04:02:15.8039893Z tmpfs           3.4G     0  3.4G   0% /sys/fs/cgroup
2020-11-12T04:02:15.8040222Z shm              64M     0   64M   0% /dev/shm
2020-11-12T04:02:15.8040517Z /dev/sdb1        84G   75G  8.9G  90% /__w
2020-11-12T04:02:15.8040859Z tmpfs           696M  768K  695M   1% /run/docker.sock
2020-11-12T04:02:15.8041228Z tmpfs           3.4G     0  3.4G   0% /proc/acpi
2020-11-12T04:02:15.8041550Z tmpfs           3.4G     0  3.4G   0% /proc/scsi
2020-11-12T04:02:15.8041886Z tmpfs           3.4G     0  3.4G   0% /sys/firmware
```
We'll have to wait until they fix that, again, I guess?

Craigacp commented on tensorflow-framework/src/main/java/org/tensorflow/framework/utils/CastHelper.java:
```diff
@@ -0,0 +1,43 @@
+   * @return the value cast to the required data type.
+   */
+  @SuppressWarnings("unchecked")
+  public static <T extends TType, U extends TType> Operand<T> cast(
```
We should open an issue to insert these cast checks into the optimizers for uniformity.

JimClarke5 replied:
I could do it in the #106  Learning Rate PR if that works.

Craigacp replied:
No, let's not hold anything up for it, it's just something to clean up later.

Craigacp commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/BinaryCrossentropy.java:
```diff
@@ -0,0 +1,230 @@
+   * @param fromLogits Whether to interpret predictions as a tensor of logit values
+   * @param labelSmoothing A number in the range, [0, 1]. When 0, no smoothing occurs. When &gt; 0,
+   *     compute the loss between the predicted labels and a smoothed version of the true labels,
+   *     where the smoothing squeezes the labels towards 0.5. Larger values of label_smoothing
```
`label_smoothing` -> `labelSmoothing`, here and elsewhere in this file.

JimClarke5 replied:
OK

Craigacp commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/CategoricalCrossentropy.java:
```diff
@@ -0,0 +1,270 @@
+   * @param fromLogits Whether to interpret predictions as a tensor of logit values
+   * @param labelSmoothing Float in [0, 1]. When 0, no smoothing occurs. When > 0, we compute the
+   *     loss between the predicted labels and a smoothed version of the true labels, where the
+   *     smoothing squeezes the labels towards 0.5. Larger values of label_smoothing correspond to
```
This one's still got the doc from BinaryCrossEntropy wrt `label_smoothing`. And it's snake_case.

JimClarke5 replied:
OK

Craigacp commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/CategoricalCrossentropy.java:
```diff
@@ -0,0 +1,270 @@
+   * @param tf the TensorFlow Ops
+   * @param fromLogits Whether to interpret predictions as a tensor of logit values
+   * @param labelSmoothing Float in [0, 1]. When 0, no smoothing occurs. When > 0, we compute the
+   *     loss between the predicted labels and a smoothed version of the true labels, where the
```
Incorrect doc.

JimClarke5 replied:
OK

Craigacp commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/CategoricalCrossentropy.java:
```diff
@@ -0,0 +1,270 @@
+   * @param fromLogits Whether to interpret predictions as a tensor of logit values
+   * @param labelSmoothing Float in [0, 1]. When 0, no smoothing occurs. When > 0, we compute the
+   *     loss between the predicted labels and a smoothed version of the true labels, where the
+   *     smoothing squeezes the labels towards 0.5. Larger values of label_smoothing correspond to
```
Incorrect doc.

JimClarke5 replied:
OK

Craigacp commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java:
```diff
@@ -0,0 +1,707 @@
+   * @param <T> the data type of the labels
+   * @return the smoothed binary labels
+   */
+  private static <T extends TNumber> Operand<T> smoothLabelsBinaryX(
```
I think this would be better called `smoothBinaryLabels` as it's not specific to the binary cross entropy as far as I can tell. But it's a private method so it's not too much of an issue.

JimClarke5 replied:
OK

Craigacp commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java:
```diff
@@ -0,0 +1,707 @@
+   * @param <T> the data type of the labels
+   * @return the smoothed categorical labels
+   */
+  private static <T extends TNumber> Operand<T> smoothLabelsCatX(
```
Similar comment to above, but `smoothCategoricalLabels`. Also I think the doc should explicitly state that it's smoothing the labels towards `1/n` where `n` is the number of classes.

JimClarke5 replied:
OK

Craigacp reviewed (changes requested):
A few more documentation things.

JimClarke5 commented:
@Craigacp I have modified the JavaDoc for `Losses.cosineSimilarity` to insert this section in front to clarify the differences between the mathematical definition of cosine similarity and this loss function.
```
   * <p>Note that it is a number between <code>-1</code> and <code>1</code>, which is different from the mathematical definition of cosine
   * similarity where <code>1</code> represents similar vectors, and <code>0</code> represents dissimilar vectors.
   * In this function, the numbers are inverted in a range of <code>-1</code> to <code>1</code>.
....
```

Craigacp commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/CategoricalCrossentropy.java:
```diff
@@ -184,10 +184,9 @@ public CategoricalCrossentropy(Ops tf, String name, boolean fromLogits, float la
-   *     heavier smoothing.
+   * @param labelSmoothing Float in <code>[0, 1]</code>. When <code>&gt; 0</code>, label values are smoothed, meaning the
+   *    confidence on label values are relaxed. e.g. <code>label_smoothing=0.2<code> means that we will use a
+   *    alue of </code>0.1<code> for label </code>0<code> and </code>0.9<code> for label </code>1<code>
```
`alue` -> `value`

JimClarke5 replied:
OK

Craigacp commented on tensorflow-framework/src/main/java/org/tensorflow/framework/losses/CategoricalCrossentropy.java:
```diff
@@ -169,10 +170,9 @@ public CategoricalCrossentropy(Ops tf, boolean fromLogits, float labelSmoothing)
-   *     smoothing squeezes the labels towards 0.5. Larger values of label_smoothing correspond to
-   *     heavier smoothing.
+   * @param labelSmoothing Float in <code>[0, 1]</code>. When <code>&gt; 0</code>, label values are smoothed, meaning the
+   *    confidence on label values are relaxed. e.g. <code>label_smoothing=0.2<code> means that we will use a
```
`label_smoothing` -> `labelSmoothing`, and in the docs below.

karllessard commented:
@Craigacp , it looks like @JimClarke5 pushed all commits for the last changes you've requested, I let you validate and dismiss your review if that's the case, thanks

@JimClarke5 , some of your unit test files don't have a valid source header, can you please add one? Also, I've noticed that now you are authoring your work to Oracle, that is ok but I just want to validate with you if it was intentional, thanks

JimClarke5 commented:
I have fixed the copyright issues. The attribution to Oracle was a mistake, as I copied the copyright from Optimizers to the Loss classes. I have replaced them with the `TensorFlow Authors` copyright.
@Craigacp did you intend this for Optimizers?

Craigacp commented:
@JimClarke5 yes, I'm required to put the Oracle copyright header on substantive external open source contributions that are part of my job. There's an internal review process for all the things that I write that are longer than a line or two.

@karllessard I've resolved my two comments, I think this is good to be merged now. I can approve it if you want.

karllessard commented:
All right, this one is merged now, thanks for great contribution again @JimClarke5 !

deansher commented:
::party::

On Tue, Nov 17, 2020 at 7:09 AM Karl Lessard <notifications@github.com>
wrote:

> All right, this one is merged now, thanks for great contribution again
> @JimClarke5 <https://github.com/JimClarke5> !
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/java/pull/129#issuecomment-728885532>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABZ7X44NRW56MHAI4DLBCDSQJRZBANCNFSM4SJLM76A>
> .
>
